src/main_menu.cpp

```diff
@@ -851,145 +856,154 @@ void main_menu::world_tab()
+                    } else if( sel3 == 1 ) { // Reset World
                         if( query_yn( _( "Remove all saves and regenerate world?" ) ) ) {
                             query_yes = true;
                             do_delete = false;
                         }
                     }
 
-                    layer = 2; // Go to world submenu, not list of worlds
                     if( query_yes ) {
-                        world_generator->delete_world( all_worldnames[sel3], do_delete );
+                        layer = 2; // Go to world submenu, not list of worlds
+
+                        world_generator->delete_world( all_worldnames[sel2 - 1], do_delete );
 
                         savegames.clear();
                         MAPBUFFER.reset();
                         overmap_buffer.clear();
 
-                        if( all_worldnames.empty() ) {
+                        if( do_delete ) {
                             sel2 = 0; // reset to create world selection
                         }
                     }
                 }
-                print_menu( w_open, sel1, iMenuOffsetX, iMenuOffsetY );
+            }
+        } else if( layer == 2 ) { // Show world names
+            if( MAP_SHARING::isSharing() && !MAP_SHARING::isWorldmenu() && !MAP_SHARING::isAdmin() ) {
+                layer = 1;
+                popup( _( "Only the admin can change worlds." ) );
+                continue;
+            }
+
+            mvwprintz( w_open, iMenuOffsetY - 2, 25 + iMenuOffsetX + extra_w / 2,
+                       ( sel2 == 0 ? h_white : c_white ), _( "Create World" ) );
+
+            int i = 1;
+            const auto all_worldnames = world_generator->all_worldnames();
+            for( auto it = all_worldnames.begin(); it != all_worldnames.end(); ++it, i++ ) {
+                int savegames_count = world_generator->get_world( *it )->world_saves.size();
+                int line = iMenuOffsetY - 2 - i;
+                nc_color color1, color2;
+                if( *it == "TUTORIAL" || *it == "DEFENSE" ) {
+                    color1 = c_ltcyan;
+                    color2 = h_ltcyan;
+                } else {
+                    color1 = c_white;
+                    color2 = h_white;
+                }
+                mvwprintz( w_open, line, 25 + iMenuOffsetX + extra_w / 2,
+                           ( sel2 == i ? color2 : color1 ), "%s (%d)", ( *it ).c_str(), savegames_count );
+            }
+
+            wrefresh( w_open );
+            refresh();
+            std::string action = ctxt.handle_input();
+
+            if( action == "DOWN" ) {
+                if( sel2 > 0 ) {
+                    --sel2;
+                } else {
+                    sel2 = all_worldnames.size();
+                }
+            } else if( action == "UP" ) {
+                if( sel2 < ( int )all_worldnames.size() ) {
+                    ++sel2;
+                } else {
+                    sel2 = 0;
+                }
```

Comment:
You repasted that block of code multiple times. This specific kind of copypaste code is also used in many places in game code.
It would be good to have a function for it, like what `clamp` does for clamping.
Say, `void wrap( int &x, int max )` or `int wrap( int x, int max )`.